{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "import primer3\n",
    "import numpy as np\n",
    "import sys\n",
    "import drawSvg as draw\n",
    "from IPython.display import display"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Primer():\n",
    "    \"\"\"\n",
    "    Class to hold all information regarding a specific primer\n",
    "    \n",
    "    Args:\n",
    "    - seq (str) - string containing the primer sequence to be used\n",
    "    - dna_conc (float) - concentration (nM) of oligo to use for Tm analysis, default is 250\n",
    "    - mv_conc (float) - concentration (mM) of monovalent cations to use for Tm analysis, default is 50\n",
    "    - dv_conc (float) - concentration (mM) of divalent cations to use for Tm analysis, default is 0\n",
    "    - dntp_conc (float) - concentration (mM) of dNTPs to use for Tm analysis, default is 0\n",
    "    \"\"\"\n",
    "    \n",
    "    def __init__(self, seq, dna_conc=250, mv_conc=50, dv_conc=0, dntp_conc=0):\n",
    "        # Convert the sequence to uppercase\n",
    "        seq = seq.upper()\n",
    "        \n",
    "        # Set the sequence to an instance variable\n",
    "        self.seq = seq\n",
    "        \n",
    "        # Set the length variable\n",
    "        self.length = len(seq)\n",
    "        \n",
    "        # Set the standard IDT OligoAnalyzer settings for analysis by primer3\n",
    "        self.thermo_settings = primer3.thermoanalysis.ThermoAnalysis(dna_conc, mv_conc, dv_conc, dntp_conc)\n",
    "        \n",
    "        # Get the Tm of the oligo\n",
    "        self.Tm = self.thermo_settings.calcTm(self.seq)\n",
    "        \n",
    "        # Get hairpin and homodimer properties\n",
    "        self.hairpin()\n",
    "        self.homodimer()\n",
    "        \n",
    "    def hairpin(self):\n",
    "        \"\"\"\n",
    "        Calculate hairpin properties of the oligo sequence\n",
    "        \"\"\"\n",
    "        hairpin_result = self.thermo_settings.calcHairpin(self.seq)\n",
    "        self.hairpin = hairpin_result.structure_found\n",
    "        self.hairpin_Tm = hairpin_result.tm\n",
    "        \n",
    "    def homodimer(self):\n",
    "        \"\"\"\n",
    "        Calculate homodimer properties of the oligo sequence\n",
    "        \"\"\"\n",
    "        homodimer_result = self.thermo_settings.calcHomodimer(self.seq)\n",
    "        self.homodimer = homodimer_result.structure_found\n",
    "        self.homodimer_Tm = homodimer_result.tm\n",
    "        \n",
    "    def rc(self):\n",
    "        \"\"\"\n",
    "        Get the reverse complement of the primer sequence\n",
    "        \"\"\"\n",
    "        comp = {\n",
    "            'A': 'T',\n",
    "            'C': 'G',\n",
    "            'G': 'C',\n",
    "            'T': 'A'\n",
    "        }\n",
    "        \n",
    "        self.seq = \"\".join([comp[nt] for nt in self.seq[::-1]])\n",
    "        \n",
    "    def report(self):\n",
    "        \"\"\"\n",
    "        Return the details of a given primer\n",
    "        \"\"\"\n",
    "        \n",
    "        details = OrderedDict({\n",
    "            'length': len(self.seq),\n",
    "            'Tm': round(self.Tm, 2),\n",
    "            'start': self.start,\n",
    "            'end': self.end,\n",
    "            'hairpin_Tm': round(self.hairpin_Tm, 2),\n",
    "            'homodimer_Tm': round(self.homodimer_Tm, 2)\n",
    "        })\n",
    "        \n",
    "        return details\n",
    "    \n",
    "    def __repr__(self):\n",
    "        return 'Seq: {}; Tm: {:.2f}'.format(self.seq, self.Tm)\n",
    "    \n",
    "    def __str__(self):\n",
    "        return self.__repr__()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Oligo():\n",
    "    \"\"\"\n",
    "    Class to hold all information regarding a specific oligo in a REDI library\n",
    "    \n",
    "    Args:\n",
    "    - seq (str) - string containing the primer sequence to be used\n",
    "    - dna_conc (float) - concentration (nM) of oligo to use for Tm analysis, default is 250\n",
    "    - mv_conc (float) - concentration (mM) of monovalent cations to use for Tm analysis, default is 50\n",
    "    - dv_conc (float) - concentration (mM) of divalent cations to use for Tm analysis, default is 0\n",
    "    - dntp_conc (float) - concentration (mM) of dNTPs to use for Tm analysis, default is 0\n",
    "    \"\"\"\n",
    "    \n",
    "    def __init__(self, template_seq, upstream_primer, downstream_primer):\n",
    "        # Define upstream and downstream primers\n",
    "        self.upstream_primer = upstream_primer\n",
    "        self.downstream_primer = downstream_primer\n",
    "        \n",
    "        # Define start and end sites, as well as the sequence of the oligo\n",
    "        self.start = self.upstream_primer.start\n",
    "        self.end = self.downstream_primer.end\n",
    "        self.seq = template_seq[self.start:self.end]\n",
    "        \n",
    "        # Define the window of interest for each oligo\n",
    "        self.window_start = self.upstream_primer.end\n",
    "        self.window_end = self.downstream_primer.start\n",
    "        self.window_seq = template_seq[self.window_start:self.window_end]\n",
    "        \n",
    "    def __repr__(self):\n",
    "        return '<Oligo ({}:{}-{}:{}) F: {:.2f}C, R: {:.2f}C>'.format(self.start, self.window_start,\n",
    "                                                                     self.window_end, self.end, self.upstream_primer.Tm,\n",
    "                                                                     self.downstream_primer.Tm)\n",
    "    \n",
    "    def __str__(self):\n",
    "        return self.__repr__()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "class LibDiagram():\n",
    "    \"\"\"\n",
    "    Class to draw SVG library diagrams\n",
    "    \n",
    "    Args:\n",
    "    - lib (Library) - Library object from which to draw potential library of interest\n",
    "    - lib_index (int) - index of library to draw\n",
    "    - width (int) - Width of the diagram in pixels\n",
    "    - height (int) - Height of the diagram in pixels\n",
    "    \"\"\"\n",
    "    \n",
    "    def __init__(self, lib, lib_index, width = 400, height = 200):\n",
    "        self.lib = lib\n",
    "        self.oligos = lib.libraries[lib_index]\n",
    "        \n",
    "        # Calculate DNA draw height and buffer distance\n",
    "        self.dna_h = height / (1.1*(len(self.oligos)+1) - 0.1)\n",
    "        self.buffer = 0.1 * self.dna_h\n",
    "        \n",
    "        self.width = width\n",
    "        self.height = height\n",
    "        \n",
    "        # Draw the diagram\n",
    "        self.draw_background()\n",
    "        self.draw_template()\n",
    "        self.draw_lib()\n",
    "        \n",
    "    def scale_x(self, nt_width):\n",
    "        \"\"\"\n",
    "        Convert nucleotide scale to pixel scale\n",
    "        \n",
    "        Args:\n",
    "        - nt_width (int) - Nucleotide width to scale to pixels\n",
    "        \n",
    "        Returns:\n",
    "        - Pixel width of the object of interest\n",
    "        \"\"\"\n",
    "        \n",
    "        # Perform conversion\n",
    "        construct_length = len(self.lib.construct)\n",
    "        nt_px = construct_length / self.width\n",
    "        return (nt_width / nt_px)\n",
    "    \n",
    "    def scale_y(self, obj_index):\n",
    "        \"\"\"\n",
    "        Convert object index to pixel scale\n",
    "        \n",
    "        Args:\n",
    "        - obj_index (int) - Object number in diagram (determines vertical position)\n",
    "        \n",
    "        Returns:\n",
    "        - Pixel y-displacement of the object of interest\n",
    "        \"\"\"\n",
    "        \n",
    "        # Offset the object index by 1 to account for the template diagram\n",
    "        offset = obj_index + 1\n",
    "        y_offset = (offset * self.dna_h) + (self.buffer * offset)\n",
    "        return(y_offset)\n",
    "        \n",
    "    def draw_background(self):\n",
    "        \"\"\"\n",
    "        Draw the blank canvas\n",
    "        \"\"\"\n",
    "        \n",
    "        # Create canvas\n",
    "        self.d = draw.Drawing(self.width, self.height)\n",
    "\n",
    "    def draw_template(self):\n",
    "        \"\"\"\n",
    "        Draw the template strand, with the coding sequence highlighted in blue\n",
    "        \"\"\"\n",
    "        \n",
    "        # Draw template\n",
    "        self.d.append(draw.Rectangle(0, 0, self.scale_x(len(self.lib.construct)),\n",
    "                                     self.dna_h, fill='#ADD8E6'))\n",
    "        self.d.append(draw.Rectangle(self.scale_x(self.lib.start), 0,\n",
    "                                     self.scale_x(len(self.lib.roi)), self.dna_h, fill='#0000CC'))\n",
    "        \n",
    "    def draw_oligo(self, oligo, y_offset):\n",
    "        \"\"\"\n",
    "        Draw an individual oligo\n",
    "        \"\"\"\n",
    "        \n",
    "        # Draw the primers\n",
    "        primer1 = draw.Rectangle(self.scale_x(oligo.start), y_offset,\n",
    "                                 self.scale_x(oligo.window_start - oligo.start), self.dna_h, fill='#FFA500')\n",
    "        primer2 = draw.Rectangle(self.scale_x(oligo.window_end), y_offset,\n",
    "                                 self.scale_x(oligo.end - oligo.window_end), self.dna_h, fill='#FFA500')\n",
    "        self.d.append(primer1)\n",
    "        self.d.append(primer2)\n",
    "        \n",
    "        # Draw spanning oligo\n",
    "        if oligo.window_start < self.lib.start and oligo.window_end > self.lib.end:\n",
    "            window1 = draw.Rectangle(self.scale_x(oligo.window_start), y_offset,\n",
    "                                     self.scale_x(self.lib.start - oligo.window_start), self.dna_h, fill='#ADD8E6')\n",
    "            window2 = draw.Rectangle(self.scale_x(self.lib.start), y_offset,\n",
    "                                     self.scale_x(self.lib.end - self.lib.start), self.dna_h, fill='#0000CC')\n",
    "            window3 = draw.Rectangle(self.scale_x(self.lib.end), y_offset,\n",
    "                                     self.scale_x(oligo.window_end - self.lib.end), self.dna_h, fill='#ADD8E6')\n",
    "            self.d.append(window1)\n",
    "            self.d.append(window2)\n",
    "            self.d.append(window3)\n",
    "            \n",
    "        # Draw upstream oligo\n",
    "        elif oligo.window_start < self.lib.start:\n",
    "            window1 = draw.Rectangle(self.scale_x(oligo.window_start), y_offset,\n",
    "                                     self.scale_x(self.lib.start - oligo.window_start), self.dna_h, fill='#ADD8E6')\n",
    "            window2 = draw.Rectangle(self.scale_x(self.lib.start), y_offset,\n",
    "                                     self.scale_x(oligo.window_end - self.lib.start), self.dna_h, fill='#0000CC')\n",
    "            self.d.append(window1)\n",
    "            self.d.append(window2)\n",
    "            \n",
    "        # Draw downstream oligo \n",
    "        elif oligo.window_end > self.lib.end:\n",
    "            window1 = draw.Rectangle(self.scale_x(oligo.window_start), y_offset,\n",
    "                                     self.scale_x(self.lib.end - oligo.window_start), self.dna_h, fill='#0000CC')\n",
    "            window2 = draw.Rectangle(self.scale_x(self.lib.end), y_offset,\n",
    "                                     self.scale_x(oligo.window_end - self.lib.end), self.dna_h, fill='#ADD8E6')\n",
    "            self.d.append(window1)\n",
    "            self.d.append(window2)\n",
    "            \n",
    "        # Draw internal oligo\n",
    "        else:\n",
    "            window = draw.Rectangle(self.scale_x(oligo.window_start), y_offset,\n",
    "                                    self.scale_x(oligo.window_end - oligo.window_start), self.dna_h, fill='#0000CC')\n",
    "            self.d.append(window)\n",
    "       \n",
    "    def draw_lib(self):\n",
    "        \"\"\"\n",
    "        Draw all of the oligos in the library of interest\n",
    "        \"\"\"\n",
    "        \n",
    "        # Loop through and draw oligos\n",
    "        for i, oligo in enumerate(self.oligos):\n",
    "            self.draw_oligo(oligo, self.scale_y(i))\n",
    "        \n",
    "    def plot(self):\n",
    "        \"\"\"\n",
    "        Display the oligo plot in a Jupyter notebook\n",
    "        \"\"\"\n",
    "        \n",
    "        display(self.d)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Library():\n",
    "    \"\"\"\n",
    "    Class to manage design and optimization of oligo tiling libraries.\n",
    "    \n",
    "    Args:\n",
    "    - fasta (str) - Path to input fasta file\n",
    "    - start (int) - Starting position of the region of interest (inclusive)\n",
    "    - end (int) - End position of the region of interest (exclusive)\n",
    "    - oligo_size (int) - Length, in basepairs, of synthesized oligos\n",
    "    - primer_min_length (int) - Minimum primer length (default = 16)\n",
    "    - primer_max_length (int) - Maximum primer length (default = 30)\n",
    "    - primer_min_tm (float) - Minimum primer Tm (default = 58)\n",
    "    - primer_max_tm (float) - Maximum primer Tm (default = 62)\n",
    "    - primer_target_tm (float) - Target primer Tm (default = 60)\n",
    "    \"\"\"\n",
    "    \n",
    "    def __init__(self, fasta, start, end, oligo_size, primer_min_length=16, primer_max_length=30,\n",
    "                 primer_min_tm=58, primer_max_tm=62, primer_target_tm=60, hairpin_tolerance=10,\n",
    "                 homodimer_tolerance=10):\n",
    "        \n",
    "        # Define instance variables\n",
    "        self.start = start\n",
    "        self.end = end\n",
    "        self.primer_min_length = primer_min_length\n",
    "        self.primer_max_length = primer_max_length\n",
    "        self.primer_min_tm = primer_min_tm\n",
    "        self.primer_max_tm = primer_max_tm\n",
    "        self.primer_target_tm = primer_target_tm\n",
    "        self.oligo_size = oligo_size\n",
    "        self.hairpin_tolerance = hairpin_tolerance\n",
    "        self.homodimer_tolerance = homodimer_tolerance\n",
    "        \n",
    "        # Read in the data\n",
    "        self.construct = list(SeqIO.parse(fasta, 'fasta'))[0]\n",
    "        \n",
    "        # Extract the region of interest\n",
    "        self.roi = self.construct[start:end]\n",
    "        \n",
    "        # Define possible breakpoints (all codon boundaries plus nucleotides beyond the end of the ROI)\n",
    "        self.breakpoints = [x + self.start for x in range(end - start) if x % 3 == 0] + [i for i in range(self.end, len(self.construct))]\n",
    "        \n",
    "        # Make a list of the oligos\n",
    "        self.libraries = []\n",
    "\n",
    "    def make_primer_with_tm(self, start_pos, rc=False):\n",
    "        \"\"\"\n",
    "        Design a primer with the target Tm.\n",
    "        \n",
    "        Args:\n",
    "        - start_pos (int) - Starting position of the primer\n",
    "        \n",
    "        Returns:\n",
    "        - Primer object with properties closest to desired Tm\n",
    "        \"\"\"\n",
    "        \n",
    "        # Get the construct sequence\n",
    "        seq = self.construct\n",
    "        \n",
    "        # Set the initial Tm to 0 and the nucleotide distance to 0\n",
    "        primer_tm = -np.inf\n",
    "        n = self.primer_min_length\n",
    "        current_diff = abs(primer_tm - self.primer_target_tm)\n",
    "        \n",
    "        # Make empty lists\n",
    "        primers = []\n",
    "        primer_tms = []\n",
    "\n",
    "        # Keep extending until the target Tm threshold is reached\n",
    "        while n <= self.primer_max_length:\n",
    "\n",
    "            # Add a nucleotide in length\n",
    "            n += 1\n",
    "\n",
    "            # Get the primer\n",
    "            primer_end = start_pos + n\n",
    "            primer_site = str(seq[start_pos:primer_end].seq)\n",
    "            primer = Primer(primer_site)\n",
    "            primers.append(primer)\n",
    "\n",
    "            # Reverse complement if need reverse primer\n",
    "            if rc:\n",
    "                primer.rc()\n",
    "\n",
    "            # Add start and end position data\n",
    "            primer.start = start_pos\n",
    "            primer.end = primer_end\n",
    "\n",
    "            # Extract the Tm\n",
    "            primer_tm = primer.Tm\n",
    "            primer_tms.append((primer_tm - self.primer_target_tm)**2)\n",
    "                \n",
    "        return(primers[np.argmin(primer_tms)])\n",
    "        \n",
    "    def get_first_oligos(self):\n",
    "        \"\"\"\n",
    "        Get the starting point oligos for the library.\n",
    "        \n",
    "        Returns:\n",
    "        - List of potential first oligos to construct a library around\n",
    "        \"\"\"\n",
    "        \n",
    "        # Initialize empty list of valid starting oligos\n",
    "        start_oligos = []\n",
    "        \n",
    "        # Loop through potential breakpoints\n",
    "        for bp in self.breakpoints[1:]:\n",
    "            \n",
    "            # Define the end of the window at the breakpoint\n",
    "            window_end = bp\n",
    "            \n",
    "            # Design the downstream primer\n",
    "            downstream_primer = self.make_primer_with_tm(window_end, rc=True)\n",
    "            \n",
    "            # Downstream primer defines the oligo position\n",
    "            oligo_end = downstream_primer.end\n",
    "            oligo_start = oligo_end - self.oligo_size\n",
    "            \n",
    "            # Make sure the oligo does not start before the beginning of the construct\n",
    "            if oligo_start < 0:\n",
    "                continue\n",
    "                \n",
    "            # Define the upstream primer\n",
    "            upstream_primer = self.make_primer_with_tm(oligo_start, rc = True)\n",
    "            \n",
    "            # Make sure both primers are of vald length\n",
    "            up_valid_length = self.primer_min_length <= upstream_primer.length <= self.primer_max_length\n",
    "            dp_valid_length = self.primer_min_length <= upstream_primer.length <= self.primer_max_length\n",
    "            \n",
    "            if not up_valid_length or not dp_valid_length:\n",
    "                continue\n",
    "            \n",
    "            # Stop search once the primers encroach on the start of the ROI\n",
    "            if upstream_primer.end > self.start:\n",
    "                break\n",
    "              \n",
    "            # If everything checks out, define an oligo and add it to the list \n",
    "            if self.primer_in_tm_range(upstream_primer) and self.primer_in_tm_range(downstream_primer):\n",
    "                oligo = Oligo(self.construct, upstream_primer, downstream_primer)\n",
    "                start_oligos.append(oligo)\n",
    "                \n",
    "        return start_oligos\n",
    "    \n",
    "    def define_primers(self, oligo_start, window_start, window_end, oligo_end):\n",
    "        \"\"\"\n",
    "        Defines primer sites based on oligo and window sizes.\n",
    "        \n",
    "        Args:\n",
    "        - oligo_start (int) - Start poition of the oligo\n",
    "        - window_start (int) - Start poition of the window\n",
    "        - window_end (int) - End poition of the window\n",
    "        - oligo_end (int) - End poition of the oligo\n",
    "        \n",
    "        Returns:\n",
    "        - Tuple of Primer objects corresponding to upstream and downstream primers\n",
    "        \"\"\"\n",
    "        \n",
    "        # Define the upstream primer\n",
    "        upstream_primer = Primer(str(self.construct[oligo_start:window_start].seq))\n",
    "        upstream_primer.start = oligo_start\n",
    "        upstream_primer.end = window_start\n",
    "        \n",
    "        # Define the downstream primer\n",
    "        downstream_primer = Primer(str(self.construct[window_end:oligo_end].seq))\n",
    "        downstream_primer.start = window_end\n",
    "        downstream_primer.end = oligo_end\n",
    "        \n",
    "        return upstream_primer, downstream_primer\n",
    "    \n",
    "    def primer_in_tm_range(self, primer):\n",
    "        \"\"\"\n",
    "        Check if primer Tm is in desired range.\n",
    "        \n",
    "        Args:\n",
    "        - primer (Primer) - Primer to test\n",
    "        \n",
    "        Returns:\n",
    "        - Boolean value whether primer is in desired Tm range\n",
    "        \"\"\"\n",
    "        return self.primer_min_tm <= primer.Tm <= self.primer_max_tm\n",
    "\n",
    "    def create_next_oligo(self, oligos):\n",
    "        \"\"\"\n",
    "        Create the next oligo in a path and set add all valid libraries to the `libraries` instance variable.\n",
    "        \n",
    "        Args:\n",
    "        - oligos (list) - List of oligos corresponding to path up to current point\n",
    "        \"\"\"\n",
    "        # Get the last oligo\n",
    "        last_oligo = oligos[-1]\n",
    "        \n",
    "        # Define the current window start\n",
    "        window_start = last_oligo.window_end\n",
    "        \n",
    "        # Loop through all of the breakpoints\n",
    "        for window_end in self.breakpoints:\n",
    "            \n",
    "            # Get the window parameters\n",
    "            window_size = window_end - window_start\n",
    "            valid_window = window_size > 0 \n",
    "            valid_size = self.primer_min_length <= ((self.oligo_size - window_size) / 2) <= self.primer_max_length\n",
    "            \n",
    "            # Get the oligo parameters\n",
    "            oligo_end = window_end + self.primer_min_length\n",
    "            oligo_start = oligo_end - self.oligo_size\n",
    "            \n",
    "            # Check if the window is valid\n",
    "            if valid_window and valid_size:\n",
    "                \n",
    "                # Get the length of the upstream primer\n",
    "                up_length = window_start - oligo_start\n",
    "            \n",
    "                # Perform sliding window search until upstream primer is too short\n",
    "                while up_length >= self.primer_min_length:\n",
    "                    \n",
    "                    # Define the primers\n",
    "                    upstream_primer, downstream_primer = self.define_primers(oligo_start, window_start, window_end, oligo_end)\n",
    "                    \n",
    "                    # Make sure primers are valid length\n",
    "                    up_valid = self.primer_min_length <= window_start - oligo_start <= self.primer_max_length\n",
    "                    dp_valid = self.primer_min_length <= oligo_end - window_end <= self.primer_max_length\n",
    "\n",
    "                    # Check if primers are in the appropriate Tm range\n",
    "                    if self.primer_in_tm_range(upstream_primer) and self.primer_in_tm_range(downstream_primer) and up_valid and dp_valid:\n",
    "                        \n",
    "                        # Define the oligo and add it to the path\n",
    "                        oligo = Oligo(self.construct, upstream_primer, downstream_primer)\n",
    "                        path = oligos.copy()\n",
    "                        path.append(oligo)\n",
    "\n",
    "                        # Add valid path to list if it is the last one, otherwise, continue search along current path\n",
    "                        if window_end > self.end:\n",
    "                            self.libraries.append(path)\n",
    "                            print(\"\\tValid libraries found:\", len(self.libraries))\n",
    "                        else:\n",
    "                            self.create_next_oligo(path)\n",
    "\n",
    "                    # Continue sliding window search\n",
    "                    oligo_start += 1\n",
    "                    oligo_end += 1\n",
    "\n",
    "                    up_length = window_start - oligo_start\n",
    "                    \n",
    "                    \n",
    "    def design_lib(self):\n",
    "        \"\"\"\n",
    "        Find all valid libraries for the construct.\n",
    "        \"\"\"\n",
    "        \n",
    "        # Find all valid first oligos\n",
    "        first_oligos = self.get_first_oligos()\n",
    "        \n",
    "        # Get all complete paths\n",
    "        for i, starting_point in enumerate(first_oligos):\n",
    "            print(\"Starting oligo {} of {}\".format(i + 1, len(first_oligos)))\n",
    "            oligos = [starting_point]\n",
    "            self.create_next_oligo(oligos)\n",
    "            \n",
    "    def score_libs(self, gc_ends=1, gc_comp=2, tm=1, hairpin_tm=0.1, homodimer_tm=0.1, num_oligos=None):\n",
    "        \"\"\"\n",
    "        Score all of the libraries, find the best library, and save as the\n",
    "        `best_lib` and `best_lib_index` instance variables.\n",
    "        \n",
    "        Args:\n",
    "        - gc_ends (float) - Multiplier for GC end score (default = 1)\n",
    "        - gc_comp (float) - Multiplier for penalty for devaiting from target\n",
    "            GC content of 0.5 (default = 2)\n",
    "        - tm (float) - Multipleir for penalty for devaiting from target primer\n",
    "            Tm (default = 1)\n",
    "        - hairpin_tm (float) - Multiplier for each degree below minimum hairpin\n",
    "            tolerance (default = .1)\n",
    "        - homodimer_tm (float) - Multiplier for each degree below minimum\n",
    "            homodimer tolerance (default = .1)\n",
    "        - num_oligos (str) Set to \"min\" to search only the libraries with the\n",
    "            minimum oligo count (default = None)\n",
    "        \"\"\"\n",
    "        \n",
    "        # Set the initial best library score to infinity \n",
    "        best_lib_score = -np.inf\n",
    "        \n",
    "        # Keep track of the minimum length library\n",
    "        min_length = np.inf\n",
    "        \n",
    "        # Loop through all valid libraries\n",
    "        for i, lib in enumerate(self.libraries):\n",
    "            \n",
    "            # Get library length\n",
    "            lib_length = len(lib)\n",
    "            \n",
    "            # Loop through all of the oligos in the library\n",
    "            for oligo in lib:\n",
    "                \n",
    "                # Start with a score of 0\n",
    "                lib_score = 0\n",
    "                \n",
    "                # Loop through all of the primers in the oligo\n",
    "                for primer in [oligo.upstream_primer, oligo.downstream_primer]:\n",
    "                    \n",
    "                    # Calculate a score for each primer and add it to the oligo score\n",
    "                    gc_end_score = gc_ends * (float(primer.seq.startswith(('C', 'G'))) +\n",
    "                                              float(primer.seq.endswith(('C', 'G'))))\n",
    "                    gc_comp_score = -gc_comp * abs((primer.seq.count('G') + primer.seq.count('C') / len(primer.seq))- 0.5)\n",
    "                    tm_mean_score = -tm * abs(primer.Tm - self.primer_target_tm)\n",
    "                    hairpin_tm_score = hairpin_tm * abs(self.hairpin_tolerance - primer.hairpin_Tm)\n",
    "                    homodimer_tm_score = homodimer_tm * abs(self.homodimer_tolerance - primer.homodimer_Tm)\n",
    "\n",
    "                    lib_score += gc_end_score + gc_comp_score + tm_mean_score + hairpin_tm_score + homodimer_tm_score\n",
    "\n",
    "            # Normalize by library length\n",
    "            lib_score /= lib_length\n",
    "            \n",
    "            # Handle instance where only libraries of minimum length should be considered\n",
    "            if num_oligos == \"min\" and lib_length < min_length:\n",
    "                best_lib_score = -np.inf\n",
    "                min_length = lib_length\n",
    "            \n",
    "            # Reset best score if this lib is the best\n",
    "            if lib_score > best_lib_score:\n",
    "                best_lib_score = lib_score\n",
    "                self.best_lib_index = i\n",
    "                self.best_lib = lib\n",
    "             \n",
    "    def plot_best_lib(self):\n",
    "        \"\"\"\n",
    "        Plot the library found as the best\n",
    "        \"\"\"\n",
    "        if not hasattr(self, \"best_lib_index\"):\n",
    "            print(\"Scoring libraries using default parameters...\")\n",
    "            self.score_libs()\n",
    "        \n",
    "        LibDiagram(self, self.best_lib_index).plot()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Starting oligo 1 of 19\n",
      "Starting oligo 2 of 19\n",
      "Starting oligo 3 of 19\n",
      "Starting oligo 4 of 19\n",
      "Starting oligo 5 of 19\n",
      "\tValid libraries found: 1\n",
      "\tValid libraries found: 2\n",
      "\tValid libraries found: 3\n",
      "\tValid libraries found: 4\n",
      "Starting oligo 6 of 19\n",
      "Starting oligo 7 of 19\n",
      "Starting oligo 8 of 19\n",
      "Starting oligo 9 of 19\n",
      "Starting oligo 10 of 19\n",
      "Starting oligo 11 of 19\n",
      "Starting oligo 12 of 19\n",
      "\tValid libraries found: 5\n",
      "\tValid libraries found: 6\n",
      "\tValid libraries found: 7\n",
      "\tValid libraries found: 8\n",
      "\tValid libraries found: 9\n",
      "\tValid libraries found: 10\n",
      "\tValid libraries found: 11\n",
      "\tValid libraries found: 12\n",
      "Starting oligo 13 of 19\n",
      "Starting oligo 14 of 19\n",
      "\tValid libraries found: 13\n",
      "\tValid libraries found: 14\n",
      "Starting oligo 15 of 19\n",
      "Starting oligo 16 of 19\n",
      "Starting oligo 17 of 19\n",
      "Starting oligo 18 of 19\n",
      "Starting oligo 19 of 19\n"
     ]
    }
   ],
   "source": [
    "lib = Library('myc_dhfr_ivtt.fa', 155, 629, 230, primer_min_tm=59, primer_max_tm=61, primer_target_tm=60)\n",
    "lib.design_lib()\n",
    "lib.score_libs()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "lib = Library('pafa_ivtt.fa', 231, 1810, 230, primer_min_tm=54, primer_max_tm=56, primer_target_tm=55)\n",
    "lib.design_lib()\n",
    "lib.score_libs()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "lib.plot_best_lib()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [],
   "source": [
    "import copy\n",
    "from Bio.Data import CodonTable\n",
    "\n",
    "codon_table = CodonTable.unambiguous_dna_by_name[\"Standard\"]\n",
    "\n",
    "class MutantLib():\n",
    "    \"\"\"\n",
    "    Class to handle mutant generation from a predesigned library.\n",
    "\n",
    "    Args:\n",
    "    - lib (Library) - Library object to use as the basis for mutant library generation\n",
    "    - index (str/int) - Either `best` or an index to use for the mutant library design\n",
    "    \"\"\"\n",
    "    \n",
    "    def __init__(self, lib, output_prefix, offset=0, idx=\"best\"):\n",
    "        self.lib = lib\n",
    "        self.output_prefix = output_prefix\n",
    "        self.offset = offset\n",
    "        \n",
    "        if idx == 'best':\n",
    "            self.target_lib = self.lib.best_lib\n",
    "        elif isinstance(idx, int):\n",
    "            self.target_lib[idx]\n",
    "        else:\n",
    "            raise TypeError('Improperly formatted target library index, must be `best` or of type int')\n",
    "        \n",
    "    def format_primer(self, primer, protein_name, oligo_id, upstream, internal):\n",
    "        if upstream:\n",
    "            location = \"u\"\n",
    "        else:\n",
    "            location = \"d\"\n",
    "\n",
    "        if internal:\n",
    "            direction = \"i\"\n",
    "        else:\n",
    "            direction = \"e\"\n",
    "\n",
    "        primer.name = \"{}_o{}_{}_{}\".format(protein_name, oligo_id, location, direction)\n",
    "        primer.description = \"\"\n",
    "        \n",
    "    def export_primers(self):\n",
    "        with open(self.output_prefix + \"_primers.csv\", \"w\") as primer_file:\n",
    "            for idx, oligo in enumerate(self.lib.best_lib):\n",
    "                idx += 1\n",
    "                upstream_primer = oligo.upstream_primer\n",
    "                downstream_primer = oligo.downstream_primer\n",
    "\n",
    "                ui_start = upstream_primer.end - 60\n",
    "                ui_start = ui_start if ui_start >= 0 else 0\n",
    "                \n",
    "                di_end = downstream_primer.start + 60\n",
    "                di_end = di_end if di_end <= len(lib.construct.seq) else len(lib.construct.seq)\n",
    "                \n",
    "                upstream_internal = lib.construct[ui_start:upstream_primer.end]\n",
    "                downstream_internal = lib.construct[downstream_primer.start:di_end].reverse_complement()\n",
    "\n",
    "                protein_name = lib.construct.name\n",
    "\n",
    "                self.format_primer(upstream_internal, protein_name, idx, True, True)\n",
    "                self.format_primer(downstream_internal, protein_name, idx, False, True)\n",
    "\n",
    "                upstream_external = lib.construct[upstream_primer.start:upstream_primer.end].reverse_complement()\n",
    "                downstream_external = lib.construct[downstream_primer.start:downstream_primer.end]\n",
    "\n",
    "                self.format_primer(upstream_external, protein_name, idx, True, False)\n",
    "                self.format_primer(downstream_external, protein_name, idx, False, False)\n",
    "\n",
    "                primers = [upstream_internal, downstream_internal, upstream_external, downstream_external]\n",
    "\n",
    "\n",
    "                for primer in primers:\n",
    "                    primer_file.write(\"{},{}\\n\".format(primer.name, str(primer.seq)))\n",
    "            \n",
    "    def annotate_codons(self):\n",
    "        breaks = [x + self.lib.start for x in range(self.lib.end - self.lib.start) if x % 3 == 0]\n",
    "        \n",
    "        start = breaks[0]\n",
    "        \n",
    "        codons = {}\n",
    "        \n",
    "        for idx, breakpoint in enumerate(breaks[1:]):\n",
    "            for oligo_idx, oligo in enumerate(self.target_lib):\n",
    "                if start >= oligo.window_start and breakpoint <= oligo.window_end:\n",
    "                    break\n",
    "            codons[idx + 1] = {'start': start,\n",
    "                               'end': breakpoint,\n",
    "                               'oligo': oligo_idx,\n",
    "                               'in_oligo_start': start - oligo.start,\n",
    "                               'in_oligo_end': breakpoint - oligo.start,\n",
    "                               'codon': str(oligo.seq[start - oligo.start:breakpoint - oligo.start].seq),\n",
    "                               'residue': str(oligo.seq[start - oligo.start:breakpoint - oligo.start].translate().seq)}\n",
    "            \n",
    "            start = breakpoint\n",
    "            \n",
    "        self.codons = codons\n",
    "            \n",
    "    def mutate(self, position, residue, mutant_codon=None):\n",
    "        back_trans_tab = {\n",
    "            'A': ('GCT', 'GCC', 'GCA', 'GCG'),\n",
    "            'C': ('TGT', 'TGC'),\n",
    "            'D': ('GAT', 'GAC'),\n",
    "            'E': ('GAA', 'GAG'),\n",
    "            'F': ('TTT', 'TTC'),\n",
    "            'G': ('GGT', 'GGC', 'GGA', 'GGG'),\n",
    "            'I': ('ATT', 'ATC', 'ATA'),\n",
    "            'H': ('CAT', 'CAC'),\n",
    "            'K': ('AAA', 'AAG'),\n",
    "            'L': ('TTA', 'TTG', 'CTT', 'CTC', 'CTA', 'CTG'),\n",
    "            'M': ('ATG',),\n",
    "            'N': ('AAT', 'AAC'),\n",
    "            'P': ('CCT', 'CCC', 'CCA', 'CCG'),\n",
    "            'Q': ('CAA', 'CAG'),\n",
    "            'R': ('CGT', 'CGC', 'CGA', 'CGG', 'AGA', 'AGG'),\n",
    "            'S': ('TCT', 'TCC', 'TCA', 'TCG', 'AGT', 'AGC'),\n",
    "            'T': ('ACT', 'ACC', 'ACA', 'ACG'),\n",
    "            'V': ('GTT', 'GTC', 'GTA', 'GTG'),\n",
    "            'W': ('TGG',),\n",
    "            'Y': ('TAT', 'TAC'),\n",
    "            '*': ('TAA', 'TAG', 'TGA'),\n",
    "        }\n",
    "        \n",
    "        if position not in list(self.codon.keys()):\n",
    "            raise KeyError('Desired position to mutate is not defined in the library.')\n",
    "            \n",
    "        else:\n",
    "            poi = self.codons[position]\n",
    "            if poi['residue'] != residue:\n",
    "                oligo = copy.copy(self.target_lib[poi['oligo']])\n",
    "                \n",
    "                if mutant_codon is not None and codon_table.forward_table[mutant_codon] == residue:\n",
    "                    codon_change = mutant_codon\n",
    "                elif mutant_codon is not None and codon_table.forward_table[mutant_codon] != residue:\n",
    "                    raise ValueError('Desired codon does not match desired residue change.')\n",
    "                elif mutant_codon is None:\n",
    "                    codon_change = codon_table.back_table[residue]\n",
    "                    \n",
    "                if poi['residue'] != residue:  \n",
    "                    oligo[poi['in_oligo_start']:poi['in_oligo_end']] = codon_change\n",
    "                    \n",
    "                    ### ADD SYN MUTATION CODE HERE ###\n",
    "                    \n",
    "                    \n",
    "                    oligo.name = '{}_{}{}{}'.format(lib.construct.name, poi['residue'], position, residue)\n",
    "                    \n",
    "                    return oligo\n",
    "    \n",
    "    def scan(self, residue):\n",
    "        oligos = "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def format_primer(primer, protein_name, oligo_id, upstream, internal):\n",
    "    if upstream:\n",
    "        location = \"u\"\n",
    "    else:\n",
    "        location = \"d\"\n",
    "        \n",
    "    if internal:\n",
    "        direction = \"i\"\n",
    "    else:\n",
    "        direction = \"e\"\n",
    "        \n",
    "    primer.name = \"{}_o{}_{}_{}\".format(protein_name, oligo_id, upstream, direction)\n",
    "    primer.description = \"\"\n",
    "\n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"test.fa\", \"w\") as handle:\n",
    "    test = lib.construct\n",
    "    test.id = \"TESTER_ID\"\n",
    "    test.name = \"TESTER_NAME\"\n",
    "    test.description = \"TESTER_DESC\"\n",
    "    \n",
    "    SeqIO.write(lib.construct, handle, 'fasta')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "SeqRecord(seq=Seq('CAAATAGGGGTTCCGCGCACATTTCCCCGAAAAGTGCTAGTGGTGCTAGCCCCG...AAC', SingleLetterAlphabet()), id='TESTER_ID', name='TESTER_NAME', description='TESTER_DESC', dbxrefs=[])"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "str(lib.construct.seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "mlib = MutantLib(lib, \"test\")\n",
    "mlib.export_primers()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{1: {'start': 155,\n",
       "  'end': 158,\n",
       "  'oligo': 0,\n",
       "  'in_oligo_start': 130,\n",
       "  'in_oligo_end': 133,\n",
       "  'codon': 'ATC',\n",
       "  'residue': 'I'},\n",
       " 2: {'start': 158,\n",
       "  'end': 161,\n",
       "  'oligo': 0,\n",
       "  'in_oligo_start': 133,\n",
       "  'in_oligo_end': 136,\n",
       "  'codon': 'AGT',\n",
       "  'residue': 'S'},\n",
       " 3: {'start': 161,\n",
       "  'end': 164,\n",
       "  'oligo': 0,\n",
       "  'in_oligo_start': 136,\n",
       "  'in_oligo_end': 139,\n",
       "  'codon': 'CTG',\n",
       "  'residue': 'L'},\n",
       " 4: {'start': 164,\n",
       "  'end': 167,\n",
       "  'oligo': 0,\n",
       "  'in_oligo_start': 139,\n",
       "  'in_oligo_end': 142,\n",
       "  'codon': 'ATT',\n",
       "  'residue': 'I'},\n",
       " 5: {'start': 167,\n",
       "  'end': 170,\n",
       "  'oligo': 0,\n",
       "  'in_oligo_start': 142,\n",
       "  'in_oligo_end': 145,\n",
       "  'codon': 'GCG',\n",
       "  'residue': 'A'},\n",
       " 6: {'start': 170,\n",
       "  'end': 173,\n",
       "  'oligo': 0,\n",
       "  'in_oligo_start': 145,\n",
       "  'in_oligo_end': 148,\n",
       "  'codon': 'GCG',\n",
       "  'residue': 'A'},\n",
       " 7: {'start': 173,\n",
       "  'end': 176,\n",
       "  'oligo': 0,\n",
       "  'in_oligo_start': 148,\n",
       "  'in_oligo_end': 151,\n",
       "  'codon': 'TTA',\n",
       "  'residue': 'L'},\n",
       " 8: {'start': 176,\n",
       "  'end': 179,\n",
       "  'oligo': 0,\n",
       "  'in_oligo_start': 151,\n",
       "  'in_oligo_end': 154,\n",
       "  'codon': 'GCG',\n",
       "  'residue': 'A'},\n",
       " 9: {'start': 179,\n",
       "  'end': 182,\n",
       "  'oligo': 0,\n",
       "  'in_oligo_start': 154,\n",
       "  'in_oligo_end': 157,\n",
       "  'codon': 'GTA',\n",
       "  'residue': 'V'},\n",
       " 10: {'start': 182,\n",
       "  'end': 185,\n",
       "  'oligo': 0,\n",
       "  'in_oligo_start': 157,\n",
       "  'in_oligo_end': 160,\n",
       "  'codon': 'GAT',\n",
       "  'residue': 'D'},\n",
       " 11: {'start': 185,\n",
       "  'end': 188,\n",
       "  'oligo': 0,\n",
       "  'in_oligo_start': 160,\n",
       "  'in_oligo_end': 163,\n",
       "  'codon': 'CGC',\n",
       "  'residue': 'R'},\n",
       " 12: {'start': 188,\n",
       "  'end': 191,\n",
       "  'oligo': 0,\n",
       "  'in_oligo_start': 163,\n",
       "  'in_oligo_end': 166,\n",
       "  'codon': 'GTT',\n",
       "  'residue': 'V'},\n",
       " 13: {'start': 191,\n",
       "  'end': 194,\n",
       "  'oligo': 0,\n",
       "  'in_oligo_start': 166,\n",
       "  'in_oligo_end': 169,\n",
       "  'codon': 'ATC',\n",
       "  'residue': 'I'},\n",
       " 14: {'start': 194,\n",
       "  'end': 197,\n",
       "  'oligo': 0,\n",
       "  'in_oligo_start': 169,\n",
       "  'in_oligo_end': 172,\n",
       "  'codon': 'GGC',\n",
       "  'residue': 'G'},\n",
       " 15: {'start': 197,\n",
       "  'end': 200,\n",
       "  'oligo': 0,\n",
       "  'in_oligo_start': 172,\n",
       "  'in_oligo_end': 175,\n",
       "  'codon': 'ATG',\n",
       "  'residue': 'M'},\n",
       " 16: {'start': 200,\n",
       "  'end': 203,\n",
       "  'oligo': 0,\n",
       "  'in_oligo_start': 175,\n",
       "  'in_oligo_end': 178,\n",
       "  'codon': 'GAA',\n",
       "  'residue': 'E'},\n",
       " 17: {'start': 203,\n",
       "  'end': 206,\n",
       "  'oligo': 0,\n",
       "  'in_oligo_start': 178,\n",
       "  'in_oligo_end': 181,\n",
       "  'codon': 'AAC',\n",
       "  'residue': 'N'},\n",
       " 18: {'start': 206,\n",
       "  'end': 209,\n",
       "  'oligo': 0,\n",
       "  'in_oligo_start': 181,\n",
       "  'in_oligo_end': 184,\n",
       "  'codon': 'GCC',\n",
       "  'residue': 'A'},\n",
       " 19: {'start': 209,\n",
       "  'end': 212,\n",
       "  'oligo': 0,\n",
       "  'in_oligo_start': 184,\n",
       "  'in_oligo_end': 187,\n",
       "  'codon': 'ATG',\n",
       "  'residue': 'M'},\n",
       " 20: {'start': 212,\n",
       "  'end': 215,\n",
       "  'oligo': 0,\n",
       "  'in_oligo_start': 187,\n",
       "  'in_oligo_end': 190,\n",
       "  'codon': 'CCG',\n",
       "  'residue': 'P'},\n",
       " 21: {'start': 215,\n",
       "  'end': 218,\n",
       "  'oligo': 0,\n",
       "  'in_oligo_start': 190,\n",
       "  'in_oligo_end': 193,\n",
       "  'codon': 'TGG',\n",
       "  'residue': 'W'},\n",
       " 22: {'start': 218,\n",
       "  'end': 221,\n",
       "  'oligo': 0,\n",
       "  'in_oligo_start': 193,\n",
       "  'in_oligo_end': 196,\n",
       "  'codon': 'AAC',\n",
       "  'residue': 'N'},\n",
       " 23: {'start': 221,\n",
       "  'end': 224,\n",
       "  'oligo': 0,\n",
       "  'in_oligo_start': 196,\n",
       "  'in_oligo_end': 199,\n",
       "  'codon': 'CTG',\n",
       "  'residue': 'L'},\n",
       " 24: {'start': 224,\n",
       "  'end': 227,\n",
       "  'oligo': 0,\n",
       "  'in_oligo_start': 199,\n",
       "  'in_oligo_end': 202,\n",
       "  'codon': 'CCT',\n",
       "  'residue': 'P'},\n",
       " 25: {'start': 227,\n",
       "  'end': 230,\n",
       "  'oligo': 0,\n",
       "  'in_oligo_start': 202,\n",
       "  'in_oligo_end': 205,\n",
       "  'codon': 'GCC',\n",
       "  'residue': 'A'},\n",
       " 26: {'start': 230,\n",
       "  'end': 233,\n",
       "  'oligo': 1,\n",
       "  'in_oligo_start': 18,\n",
       "  'in_oligo_end': 21,\n",
       "  'codon': 'GAT',\n",
       "  'residue': 'D'},\n",
       " 27: {'start': 233,\n",
       "  'end': 236,\n",
       "  'oligo': 1,\n",
       "  'in_oligo_start': 21,\n",
       "  'in_oligo_end': 24,\n",
       "  'codon': 'CTC',\n",
       "  'residue': 'L'},\n",
       " 28: {'start': 236,\n",
       "  'end': 239,\n",
       "  'oligo': 1,\n",
       "  'in_oligo_start': 24,\n",
       "  'in_oligo_end': 27,\n",
       "  'codon': 'GCC',\n",
       "  'residue': 'A'},\n",
       " 29: {'start': 239,\n",
       "  'end': 242,\n",
       "  'oligo': 1,\n",
       "  'in_oligo_start': 27,\n",
       "  'in_oligo_end': 30,\n",
       "  'codon': 'TGG',\n",
       "  'residue': 'W'},\n",
       " 30: {'start': 242,\n",
       "  'end': 245,\n",
       "  'oligo': 1,\n",
       "  'in_oligo_start': 30,\n",
       "  'in_oligo_end': 33,\n",
       "  'codon': 'TTT',\n",
       "  'residue': 'F'},\n",
       " 31: {'start': 245,\n",
       "  'end': 248,\n",
       "  'oligo': 1,\n",
       "  'in_oligo_start': 33,\n",
       "  'in_oligo_end': 36,\n",
       "  'codon': 'AAA',\n",
       "  'residue': 'K'},\n",
       " 32: {'start': 248,\n",
       "  'end': 251,\n",
       "  'oligo': 1,\n",
       "  'in_oligo_start': 36,\n",
       "  'in_oligo_end': 39,\n",
       "  'codon': 'CGC',\n",
       "  'residue': 'R'},\n",
       " 33: {'start': 251,\n",
       "  'end': 254,\n",
       "  'oligo': 1,\n",
       "  'in_oligo_start': 39,\n",
       "  'in_oligo_end': 42,\n",
       "  'codon': 'AAC',\n",
       "  'residue': 'N'},\n",
       " 34: {'start': 254,\n",
       "  'end': 257,\n",
       "  'oligo': 1,\n",
       "  'in_oligo_start': 42,\n",
       "  'in_oligo_end': 45,\n",
       "  'codon': 'ACC',\n",
       "  'residue': 'T'},\n",
       " 35: {'start': 257,\n",
       "  'end': 260,\n",
       "  'oligo': 1,\n",
       "  'in_oligo_start': 45,\n",
       "  'in_oligo_end': 48,\n",
       "  'codon': 'TTA',\n",
       "  'residue': 'L'},\n",
       " 36: {'start': 260,\n",
       "  'end': 263,\n",
       "  'oligo': 1,\n",
       "  'in_oligo_start': 48,\n",
       "  'in_oligo_end': 51,\n",
       "  'codon': 'AAT',\n",
       "  'residue': 'N'},\n",
       " 37: {'start': 263,\n",
       "  'end': 266,\n",
       "  'oligo': 1,\n",
       "  'in_oligo_start': 51,\n",
       "  'in_oligo_end': 54,\n",
       "  'codon': 'AAA',\n",
       "  'residue': 'K'},\n",
       " 38: {'start': 266,\n",
       "  'end': 269,\n",
       "  'oligo': 1,\n",
       "  'in_oligo_start': 54,\n",
       "  'in_oligo_end': 57,\n",
       "  'codon': 'CCC',\n",
       "  'residue': 'P'},\n",
       " 39: {'start': 269,\n",
       "  'end': 272,\n",
       "  'oligo': 1,\n",
       "  'in_oligo_start': 57,\n",
       "  'in_oligo_end': 60,\n",
       "  'codon': 'GTG',\n",
       "  'residue': 'V'},\n",
       " 40: {'start': 272,\n",
       "  'end': 275,\n",
       "  'oligo': 1,\n",
       "  'in_oligo_start': 60,\n",
       "  'in_oligo_end': 63,\n",
       "  'codon': 'ATT',\n",
       "  'residue': 'I'},\n",
       " 41: {'start': 275,\n",
       "  'end': 278,\n",
       "  'oligo': 1,\n",
       "  'in_oligo_start': 63,\n",
       "  'in_oligo_end': 66,\n",
       "  'codon': 'ATG',\n",
       "  'residue': 'M'},\n",
       " 42: {'start': 278,\n",
       "  'end': 281,\n",
       "  'oligo': 1,\n",
       "  'in_oligo_start': 66,\n",
       "  'in_oligo_end': 69,\n",
       "  'codon': 'GGC',\n",
       "  'residue': 'G'},\n",
       " 43: {'start': 281,\n",
       "  'end': 284,\n",
       "  'oligo': 1,\n",
       "  'in_oligo_start': 69,\n",
       "  'in_oligo_end': 72,\n",
       "  'codon': 'CGC',\n",
       "  'residue': 'R'},\n",
       " 44: {'start': 284,\n",
       "  'end': 287,\n",
       "  'oligo': 1,\n",
       "  'in_oligo_start': 72,\n",
       "  'in_oligo_end': 75,\n",
       "  'codon': 'CAT',\n",
       "  'residue': 'H'},\n",
       " 45: {'start': 287,\n",
       "  'end': 290,\n",
       "  'oligo': 1,\n",
       "  'in_oligo_start': 75,\n",
       "  'in_oligo_end': 78,\n",
       "  'codon': 'ACC',\n",
       "  'residue': 'T'},\n",
       " 46: {'start': 290,\n",
       "  'end': 293,\n",
       "  'oligo': 1,\n",
       "  'in_oligo_start': 78,\n",
       "  'in_oligo_end': 81,\n",
       "  'codon': 'TGG',\n",
       "  'residue': 'W'},\n",
       " 47: {'start': 293,\n",
       "  'end': 296,\n",
       "  'oligo': 1,\n",
       "  'in_oligo_start': 81,\n",
       "  'in_oligo_end': 84,\n",
       "  'codon': 'GAA',\n",
       "  'residue': 'E'},\n",
       " 48: {'start': 296,\n",
       "  'end': 299,\n",
       "  'oligo': 1,\n",
       "  'in_oligo_start': 84,\n",
       "  'in_oligo_end': 87,\n",
       "  'codon': 'TCA',\n",
       "  'residue': 'S'},\n",
       " 49: {'start': 299,\n",
       "  'end': 302,\n",
       "  'oligo': 1,\n",
       "  'in_oligo_start': 87,\n",
       "  'in_oligo_end': 90,\n",
       "  'codon': 'ATC',\n",
       "  'residue': 'I'},\n",
       " 50: {'start': 302,\n",
       "  'end': 305,\n",
       "  'oligo': 1,\n",
       "  'in_oligo_start': 90,\n",
       "  'in_oligo_end': 93,\n",
       "  'codon': 'GGT',\n",
       "  'residue': 'G'},\n",
       " 51: {'start': 305,\n",
       "  'end': 308,\n",
       "  'oligo': 1,\n",
       "  'in_oligo_start': 93,\n",
       "  'in_oligo_end': 96,\n",
       "  'codon': 'CGT',\n",
       "  'residue': 'R'},\n",
       " 52: {'start': 308,\n",
       "  'end': 311,\n",
       "  'oligo': 1,\n",
       "  'in_oligo_start': 96,\n",
       "  'in_oligo_end': 99,\n",
       "  'codon': 'CCG',\n",
       "  'residue': 'P'},\n",
       " 53: {'start': 311,\n",
       "  'end': 314,\n",
       "  'oligo': 1,\n",
       "  'in_oligo_start': 99,\n",
       "  'in_oligo_end': 102,\n",
       "  'codon': 'TTG',\n",
       "  'residue': 'L'},\n",
       " 54: {'start': 314,\n",
       "  'end': 317,\n",
       "  'oligo': 1,\n",
       "  'in_oligo_start': 102,\n",
       "  'in_oligo_end': 105,\n",
       "  'codon': 'CCA',\n",
       "  'residue': 'P'},\n",
       " 55: {'start': 317,\n",
       "  'end': 320,\n",
       "  'oligo': 1,\n",
       "  'in_oligo_start': 105,\n",
       "  'in_oligo_end': 108,\n",
       "  'codon': 'GGA',\n",
       "  'residue': 'G'},\n",
       " 56: {'start': 320,\n",
       "  'end': 323,\n",
       "  'oligo': 1,\n",
       "  'in_oligo_start': 108,\n",
       "  'in_oligo_end': 111,\n",
       "  'codon': 'CGC',\n",
       "  'residue': 'R'},\n",
       " 57: {'start': 323,\n",
       "  'end': 326,\n",
       "  'oligo': 1,\n",
       "  'in_oligo_start': 111,\n",
       "  'in_oligo_end': 114,\n",
       "  'codon': 'AAA',\n",
       "  'residue': 'K'},\n",
       " 58: {'start': 326,\n",
       "  'end': 329,\n",
       "  'oligo': 1,\n",
       "  'in_oligo_start': 114,\n",
       "  'in_oligo_end': 117,\n",
       "  'codon': 'AAT',\n",
       "  'residue': 'N'},\n",
       " 59: {'start': 329,\n",
       "  'end': 332,\n",
       "  'oligo': 1,\n",
       "  'in_oligo_start': 117,\n",
       "  'in_oligo_end': 120,\n",
       "  'codon': 'ATT',\n",
       "  'residue': 'I'},\n",
       " 60: {'start': 332,\n",
       "  'end': 335,\n",
       "  'oligo': 1,\n",
       "  'in_oligo_start': 120,\n",
       "  'in_oligo_end': 123,\n",
       "  'codon': 'ATC',\n",
       "  'residue': 'I'},\n",
       " 61: {'start': 335,\n",
       "  'end': 338,\n",
       "  'oligo': 1,\n",
       "  'in_oligo_start': 123,\n",
       "  'in_oligo_end': 126,\n",
       "  'codon': 'CTC',\n",
       "  'residue': 'L'},\n",
       " 62: {'start': 338,\n",
       "  'end': 341,\n",
       "  'oligo': 1,\n",
       "  'in_oligo_start': 126,\n",
       "  'in_oligo_end': 129,\n",
       "  'codon': 'AGC',\n",
       "  'residue': 'S'},\n",
       " 63: {'start': 341,\n",
       "  'end': 344,\n",
       "  'oligo': 1,\n",
       "  'in_oligo_start': 129,\n",
       "  'in_oligo_end': 132,\n",
       "  'codon': 'AGT',\n",
       "  'residue': 'S'},\n",
       " 64: {'start': 344,\n",
       "  'end': 347,\n",
       "  'oligo': 1,\n",
       "  'in_oligo_start': 132,\n",
       "  'in_oligo_end': 135,\n",
       "  'codon': 'CAA',\n",
       "  'residue': 'Q'},\n",
       " 65: {'start': 347,\n",
       "  'end': 350,\n",
       "  'oligo': 1,\n",
       "  'in_oligo_start': 135,\n",
       "  'in_oligo_end': 138,\n",
       "  'codon': 'CCG',\n",
       "  'residue': 'P'},\n",
       " 66: {'start': 350,\n",
       "  'end': 353,\n",
       "  'oligo': 1,\n",
       "  'in_oligo_start': 138,\n",
       "  'in_oligo_end': 141,\n",
       "  'codon': 'GGT',\n",
       "  'residue': 'G'},\n",
       " 67: {'start': 353,\n",
       "  'end': 356,\n",
       "  'oligo': 1,\n",
       "  'in_oligo_start': 141,\n",
       "  'in_oligo_end': 144,\n",
       "  'codon': 'ACG',\n",
       "  'residue': 'T'},\n",
       " 68: {'start': 356,\n",
       "  'end': 359,\n",
       "  'oligo': 1,\n",
       "  'in_oligo_start': 144,\n",
       "  'in_oligo_end': 147,\n",
       "  'codon': 'GAC',\n",
       "  'residue': 'D'},\n",
       " 69: {'start': 359,\n",
       "  'end': 362,\n",
       "  'oligo': 1,\n",
       "  'in_oligo_start': 147,\n",
       "  'in_oligo_end': 150,\n",
       "  'codon': 'GAT',\n",
       "  'residue': 'D'},\n",
       " 70: {'start': 362,\n",
       "  'end': 365,\n",
       "  'oligo': 1,\n",
       "  'in_oligo_start': 150,\n",
       "  'in_oligo_end': 153,\n",
       "  'codon': 'CGC',\n",
       "  'residue': 'R'},\n",
       " 71: {'start': 365,\n",
       "  'end': 368,\n",
       "  'oligo': 1,\n",
       "  'in_oligo_start': 153,\n",
       "  'in_oligo_end': 156,\n",
       "  'codon': 'GTA',\n",
       "  'residue': 'V'},\n",
       " 72: {'start': 368,\n",
       "  'end': 371,\n",
       "  'oligo': 1,\n",
       "  'in_oligo_start': 156,\n",
       "  'in_oligo_end': 159,\n",
       "  'codon': 'ACG',\n",
       "  'residue': 'T'},\n",
       " 73: {'start': 371,\n",
       "  'end': 374,\n",
       "  'oligo': 1,\n",
       "  'in_oligo_start': 159,\n",
       "  'in_oligo_end': 162,\n",
       "  'codon': 'TGG',\n",
       "  'residue': 'W'},\n",
       " 74: {'start': 374,\n",
       "  'end': 377,\n",
       "  'oligo': 1,\n",
       "  'in_oligo_start': 162,\n",
       "  'in_oligo_end': 165,\n",
       "  'codon': 'GTG',\n",
       "  'residue': 'V'},\n",
       " 75: {'start': 377,\n",
       "  'end': 380,\n",
       "  'oligo': 1,\n",
       "  'in_oligo_start': 165,\n",
       "  'in_oligo_end': 168,\n",
       "  'codon': 'AAG',\n",
       "  'residue': 'K'},\n",
       " 76: {'start': 380,\n",
       "  'end': 383,\n",
       "  'oligo': 1,\n",
       "  'in_oligo_start': 168,\n",
       "  'in_oligo_end': 171,\n",
       "  'codon': 'TCG',\n",
       "  'residue': 'S'},\n",
       " 77: {'start': 383,\n",
       "  'end': 386,\n",
       "  'oligo': 1,\n",
       "  'in_oligo_start': 171,\n",
       "  'in_oligo_end': 174,\n",
       "  'codon': 'GTG',\n",
       "  'residue': 'V'},\n",
       " 78: {'start': 386,\n",
       "  'end': 389,\n",
       "  'oligo': 1,\n",
       "  'in_oligo_start': 174,\n",
       "  'in_oligo_end': 177,\n",
       "  'codon': 'GAT',\n",
       "  'residue': 'D'},\n",
       " 79: {'start': 389,\n",
       "  'end': 392,\n",
       "  'oligo': 1,\n",
       "  'in_oligo_start': 177,\n",
       "  'in_oligo_end': 180,\n",
       "  'codon': 'GAA',\n",
       "  'residue': 'E'},\n",
       " 80: {'start': 392,\n",
       "  'end': 395,\n",
       "  'oligo': 1,\n",
       "  'in_oligo_start': 180,\n",
       "  'in_oligo_end': 183,\n",
       "  'codon': 'GCC',\n",
       "  'residue': 'A'},\n",
       " 81: {'start': 395,\n",
       "  'end': 398,\n",
       "  'oligo': 1,\n",
       "  'in_oligo_start': 183,\n",
       "  'in_oligo_end': 186,\n",
       "  'codon': 'ATC',\n",
       "  'residue': 'I'},\n",
       " 82: {'start': 398,\n",
       "  'end': 401,\n",
       "  'oligo': 1,\n",
       "  'in_oligo_start': 186,\n",
       "  'in_oligo_end': 189,\n",
       "  'codon': 'GCG',\n",
       "  'residue': 'A'},\n",
       " 83: {'start': 401,\n",
       "  'end': 404,\n",
       "  'oligo': 1,\n",
       "  'in_oligo_start': 189,\n",
       "  'in_oligo_end': 192,\n",
       "  'codon': 'GCG',\n",
       "  'residue': 'A'},\n",
       " 84: {'start': 404,\n",
       "  'end': 407,\n",
       "  'oligo': 1,\n",
       "  'in_oligo_start': 192,\n",
       "  'in_oligo_end': 195,\n",
       "  'codon': 'TGT',\n",
       "  'residue': 'C'},\n",
       " 85: {'start': 407,\n",
       "  'end': 410,\n",
       "  'oligo': 1,\n",
       "  'in_oligo_start': 195,\n",
       "  'in_oligo_end': 198,\n",
       "  'codon': 'GGT',\n",
       "  'residue': 'G'},\n",
       " 86: {'start': 410,\n",
       "  'end': 413,\n",
       "  'oligo': 1,\n",
       "  'in_oligo_start': 198,\n",
       "  'in_oligo_end': 201,\n",
       "  'codon': 'GAC',\n",
       "  'residue': 'D'},\n",
       " 87: {'start': 413,\n",
       "  'end': 416,\n",
       "  'oligo': 1,\n",
       "  'in_oligo_start': 201,\n",
       "  'in_oligo_end': 204,\n",
       "  'codon': 'GTA',\n",
       "  'residue': 'V'},\n",
       " 88: {'start': 416,\n",
       "  'end': 419,\n",
       "  'oligo': 1,\n",
       "  'in_oligo_start': 204,\n",
       "  'in_oligo_end': 207,\n",
       "  'codon': 'CCA',\n",
       "  'residue': 'P'},\n",
       " 89: {'start': 419,\n",
       "  'end': 422,\n",
       "  'oligo': 1,\n",
       "  'in_oligo_start': 207,\n",
       "  'in_oligo_end': 210,\n",
       "  'codon': 'GAA',\n",
       "  'residue': 'E'},\n",
       " 90: {'start': 422,\n",
       "  'end': 425,\n",
       "  'oligo': 2,\n",
       "  'in_oligo_start': 22,\n",
       "  'in_oligo_end': 25,\n",
       "  'codon': 'ATC',\n",
       "  'residue': 'I'},\n",
       " 91: {'start': 425,\n",
       "  'end': 428,\n",
       "  'oligo': 2,\n",
       "  'in_oligo_start': 25,\n",
       "  'in_oligo_end': 28,\n",
       "  'codon': 'ATG',\n",
       "  'residue': 'M'},\n",
       " 92: {'start': 428,\n",
       "  'end': 431,\n",
       "  'oligo': 2,\n",
       "  'in_oligo_start': 28,\n",
       "  'in_oligo_end': 31,\n",
       "  'codon': 'GTG',\n",
       "  'residue': 'V'},\n",
       " 93: {'start': 431,\n",
       "  'end': 434,\n",
       "  'oligo': 2,\n",
       "  'in_oligo_start': 31,\n",
       "  'in_oligo_end': 34,\n",
       "  'codon': 'ATT',\n",
       "  'residue': 'I'},\n",
       " 94: {'start': 434,\n",
       "  'end': 437,\n",
       "  'oligo': 2,\n",
       "  'in_oligo_start': 34,\n",
       "  'in_oligo_end': 37,\n",
       "  'codon': 'GGC',\n",
       "  'residue': 'G'},\n",
       " 95: {'start': 437,\n",
       "  'end': 440,\n",
       "  'oligo': 2,\n",
       "  'in_oligo_start': 37,\n",
       "  'in_oligo_end': 40,\n",
       "  'codon': 'GGC',\n",
       "  'residue': 'G'},\n",
       " 96: {'start': 440,\n",
       "  'end': 443,\n",
       "  'oligo': 2,\n",
       "  'in_oligo_start': 40,\n",
       "  'in_oligo_end': 43,\n",
       "  'codon': 'GGT',\n",
       "  'residue': 'G'},\n",
       " 97: {'start': 443,\n",
       "  'end': 446,\n",
       "  'oligo': 2,\n",
       "  'in_oligo_start': 43,\n",
       "  'in_oligo_end': 46,\n",
       "  'codon': 'CGC',\n",
       "  'residue': 'R'},\n",
       " 98: {'start': 446,\n",
       "  'end': 449,\n",
       "  'oligo': 2,\n",
       "  'in_oligo_start': 46,\n",
       "  'in_oligo_end': 49,\n",
       "  'codon': 'GTT',\n",
       "  'residue': 'V'},\n",
       " 99: {'start': 449,\n",
       "  'end': 452,\n",
       "  'oligo': 2,\n",
       "  'in_oligo_start': 49,\n",
       "  'in_oligo_end': 52,\n",
       "  'codon': 'TAT',\n",
       "  'residue': 'Y'},\n",
       " 100: {'start': 452,\n",
       "  'end': 455,\n",
       "  'oligo': 2,\n",
       "  'in_oligo_start': 52,\n",
       "  'in_oligo_end': 55,\n",
       "  'codon': 'GAA',\n",
       "  'residue': 'E'},\n",
       " 101: {'start': 455,\n",
       "  'end': 458,\n",
       "  'oligo': 2,\n",
       "  'in_oligo_start': 55,\n",
       "  'in_oligo_end': 58,\n",
       "  'codon': 'CAG',\n",
       "  'residue': 'Q'},\n",
       " 102: {'start': 458,\n",
       "  'end': 461,\n",
       "  'oligo': 2,\n",
       "  'in_oligo_start': 58,\n",
       "  'in_oligo_end': 61,\n",
       "  'codon': 'TTC',\n",
       "  'residue': 'F'},\n",
       " 103: {'start': 461,\n",
       "  'end': 464,\n",
       "  'oligo': 2,\n",
       "  'in_oligo_start': 61,\n",
       "  'in_oligo_end': 64,\n",
       "  'codon': 'TTG',\n",
       "  'residue': 'L'},\n",
       " 104: {'start': 464,\n",
       "  'end': 467,\n",
       "  'oligo': 2,\n",
       "  'in_oligo_start': 64,\n",
       "  'in_oligo_end': 67,\n",
       "  'codon': 'CCA',\n",
       "  'residue': 'P'},\n",
       " 105: {'start': 467,\n",
       "  'end': 470,\n",
       "  'oligo': 2,\n",
       "  'in_oligo_start': 67,\n",
       "  'in_oligo_end': 70,\n",
       "  'codon': 'AAA',\n",
       "  'residue': 'K'},\n",
       " 106: {'start': 470,\n",
       "  'end': 473,\n",
       "  'oligo': 2,\n",
       "  'in_oligo_start': 70,\n",
       "  'in_oligo_end': 73,\n",
       "  'codon': 'GCG',\n",
       "  'residue': 'A'},\n",
       " 107: {'start': 473,\n",
       "  'end': 476,\n",
       "  'oligo': 2,\n",
       "  'in_oligo_start': 73,\n",
       "  'in_oligo_end': 76,\n",
       "  'codon': 'CAA',\n",
       "  'residue': 'Q'},\n",
       " 108: {'start': 476,\n",
       "  'end': 479,\n",
       "  'oligo': 2,\n",
       "  'in_oligo_start': 76,\n",
       "  'in_oligo_end': 79,\n",
       "  'codon': 'AAA',\n",
       "  'residue': 'K'},\n",
       " 109: {'start': 479,\n",
       "  'end': 482,\n",
       "  'oligo': 2,\n",
       "  'in_oligo_start': 79,\n",
       "  'in_oligo_end': 82,\n",
       "  'codon': 'CTG',\n",
       "  'residue': 'L'},\n",
       " 110: {'start': 482,\n",
       "  'end': 485,\n",
       "  'oligo': 2,\n",
       "  'in_oligo_start': 82,\n",
       "  'in_oligo_end': 85,\n",
       "  'codon': 'TAT',\n",
       "  'residue': 'Y'},\n",
       " 111: {'start': 485,\n",
       "  'end': 488,\n",
       "  'oligo': 2,\n",
       "  'in_oligo_start': 85,\n",
       "  'in_oligo_end': 88,\n",
       "  'codon': 'CTG',\n",
       "  'residue': 'L'},\n",
       " 112: {'start': 488,\n",
       "  'end': 491,\n",
       "  'oligo': 2,\n",
       "  'in_oligo_start': 88,\n",
       "  'in_oligo_end': 91,\n",
       "  'codon': 'ACG',\n",
       "  'residue': 'T'},\n",
       " 113: {'start': 491,\n",
       "  'end': 494,\n",
       "  'oligo': 2,\n",
       "  'in_oligo_start': 91,\n",
       "  'in_oligo_end': 94,\n",
       "  'codon': 'CAT',\n",
       "  'residue': 'H'},\n",
       " 114: {'start': 494,\n",
       "  'end': 497,\n",
       "  'oligo': 2,\n",
       "  'in_oligo_start': 94,\n",
       "  'in_oligo_end': 97,\n",
       "  'codon': 'ATC',\n",
       "  'residue': 'I'},\n",
       " 115: {'start': 497,\n",
       "  'end': 500,\n",
       "  'oligo': 2,\n",
       "  'in_oligo_start': 97,\n",
       "  'in_oligo_end': 100,\n",
       "  'codon': 'GAC',\n",
       "  'residue': 'D'},\n",
       " 116: {'start': 500,\n",
       "  'end': 503,\n",
       "  'oligo': 2,\n",
       "  'in_oligo_start': 100,\n",
       "  'in_oligo_end': 103,\n",
       "  'codon': 'GCA',\n",
       "  'residue': 'A'},\n",
       " 117: {'start': 503,\n",
       "  'end': 506,\n",
       "  'oligo': 2,\n",
       "  'in_oligo_start': 103,\n",
       "  'in_oligo_end': 106,\n",
       "  'codon': 'GAA',\n",
       "  'residue': 'E'},\n",
       " 118: {'start': 506,\n",
       "  'end': 509,\n",
       "  'oligo': 2,\n",
       "  'in_oligo_start': 106,\n",
       "  'in_oligo_end': 109,\n",
       "  'codon': 'GTG',\n",
       "  'residue': 'V'},\n",
       " 119: {'start': 509,\n",
       "  'end': 512,\n",
       "  'oligo': 2,\n",
       "  'in_oligo_start': 109,\n",
       "  'in_oligo_end': 112,\n",
       "  'codon': 'GAA',\n",
       "  'residue': 'E'},\n",
       " 120: {'start': 512,\n",
       "  'end': 515,\n",
       "  'oligo': 2,\n",
       "  'in_oligo_start': 112,\n",
       "  'in_oligo_end': 115,\n",
       "  'codon': 'GGC',\n",
       "  'residue': 'G'},\n",
       " 121: {'start': 515,\n",
       "  'end': 518,\n",
       "  'oligo': 2,\n",
       "  'in_oligo_start': 115,\n",
       "  'in_oligo_end': 118,\n",
       "  'codon': 'GAC',\n",
       "  'residue': 'D'},\n",
       " 122: {'start': 518,\n",
       "  'end': 521,\n",
       "  'oligo': 2,\n",
       "  'in_oligo_start': 118,\n",
       "  'in_oligo_end': 121,\n",
       "  'codon': 'ACC',\n",
       "  'residue': 'T'},\n",
       " 123: {'start': 521,\n",
       "  'end': 524,\n",
       "  'oligo': 2,\n",
       "  'in_oligo_start': 121,\n",
       "  'in_oligo_end': 124,\n",
       "  'codon': 'CAT',\n",
       "  'residue': 'H'},\n",
       " 124: {'start': 524,\n",
       "  'end': 527,\n",
       "  'oligo': 2,\n",
       "  'in_oligo_start': 124,\n",
       "  'in_oligo_end': 127,\n",
       "  'codon': 'TTC',\n",
       "  'residue': 'F'},\n",
       " 125: {'start': 527,\n",
       "  'end': 530,\n",
       "  'oligo': 2,\n",
       "  'in_oligo_start': 127,\n",
       "  'in_oligo_end': 130,\n",
       "  'codon': 'CCG',\n",
       "  'residue': 'P'},\n",
       " 126: {'start': 530,\n",
       "  'end': 533,\n",
       "  'oligo': 2,\n",
       "  'in_oligo_start': 130,\n",
       "  'in_oligo_end': 133,\n",
       "  'codon': 'GAT',\n",
       "  'residue': 'D'},\n",
       " 127: {'start': 533,\n",
       "  'end': 536,\n",
       "  'oligo': 2,\n",
       "  'in_oligo_start': 133,\n",
       "  'in_oligo_end': 136,\n",
       "  'codon': 'TAC',\n",
       "  'residue': 'Y'},\n",
       " 128: {'start': 536,\n",
       "  'end': 539,\n",
       "  'oligo': 2,\n",
       "  'in_oligo_start': 136,\n",
       "  'in_oligo_end': 139,\n",
       "  'codon': 'GAG',\n",
       "  'residue': 'E'},\n",
       " 129: {'start': 539,\n",
       "  'end': 542,\n",
       "  'oligo': 2,\n",
       "  'in_oligo_start': 139,\n",
       "  'in_oligo_end': 142,\n",
       "  'codon': 'CCG',\n",
       "  'residue': 'P'},\n",
       " 130: {'start': 542,\n",
       "  'end': 545,\n",
       "  'oligo': 2,\n",
       "  'in_oligo_start': 142,\n",
       "  'in_oligo_end': 145,\n",
       "  'codon': 'GAT',\n",
       "  'residue': 'D'},\n",
       " 131: {'start': 545,\n",
       "  'end': 548,\n",
       "  'oligo': 2,\n",
       "  'in_oligo_start': 145,\n",
       "  'in_oligo_end': 148,\n",
       "  'codon': 'GAC',\n",
       "  'residue': 'D'},\n",
       " 132: {'start': 548,\n",
       "  'end': 551,\n",
       "  'oligo': 2,\n",
       "  'in_oligo_start': 148,\n",
       "  'in_oligo_end': 151,\n",
       "  'codon': 'TGG',\n",
       "  'residue': 'W'},\n",
       " 133: {'start': 551,\n",
       "  'end': 554,\n",
       "  'oligo': 2,\n",
       "  'in_oligo_start': 151,\n",
       "  'in_oligo_end': 154,\n",
       "  'codon': 'GAA',\n",
       "  'residue': 'E'},\n",
       " 134: {'start': 554,\n",
       "  'end': 557,\n",
       "  'oligo': 2,\n",
       "  'in_oligo_start': 154,\n",
       "  'in_oligo_end': 157,\n",
       "  'codon': 'TCG',\n",
       "  'residue': 'S'},\n",
       " 135: {'start': 557,\n",
       "  'end': 560,\n",
       "  'oligo': 2,\n",
       "  'in_oligo_start': 157,\n",
       "  'in_oligo_end': 160,\n",
       "  'codon': 'GTA',\n",
       "  'residue': 'V'},\n",
       " 136: {'start': 560,\n",
       "  'end': 563,\n",
       "  'oligo': 2,\n",
       "  'in_oligo_start': 160,\n",
       "  'in_oligo_end': 163,\n",
       "  'codon': 'TTC',\n",
       "  'residue': 'F'},\n",
       " 137: {'start': 563,\n",
       "  'end': 566,\n",
       "  'oligo': 2,\n",
       "  'in_oligo_start': 163,\n",
       "  'in_oligo_end': 166,\n",
       "  'codon': 'AGC',\n",
       "  'residue': 'S'},\n",
       " 138: {'start': 566,\n",
       "  'end': 569,\n",
       "  'oligo': 2,\n",
       "  'in_oligo_start': 166,\n",
       "  'in_oligo_end': 169,\n",
       "  'codon': 'GAA',\n",
       "  'residue': 'E'},\n",
       " 139: {'start': 569,\n",
       "  'end': 572,\n",
       "  'oligo': 2,\n",
       "  'in_oligo_start': 169,\n",
       "  'in_oligo_end': 172,\n",
       "  'codon': 'TTC',\n",
       "  'residue': 'F'},\n",
       " 140: {'start': 572,\n",
       "  'end': 575,\n",
       "  'oligo': 2,\n",
       "  'in_oligo_start': 172,\n",
       "  'in_oligo_end': 175,\n",
       "  'codon': 'CAC',\n",
       "  'residue': 'H'},\n",
       " 141: {'start': 575,\n",
       "  'end': 578,\n",
       "  'oligo': 2,\n",
       "  'in_oligo_start': 175,\n",
       "  'in_oligo_end': 178,\n",
       "  'codon': 'GAT',\n",
       "  'residue': 'D'},\n",
       " 142: {'start': 578,\n",
       "  'end': 581,\n",
       "  'oligo': 2,\n",
       "  'in_oligo_start': 178,\n",
       "  'in_oligo_end': 181,\n",
       "  'codon': 'GCT',\n",
       "  'residue': 'A'},\n",
       " 143: {'start': 581,\n",
       "  'end': 584,\n",
       "  'oligo': 2,\n",
       "  'in_oligo_start': 181,\n",
       "  'in_oligo_end': 184,\n",
       "  'codon': 'GAT',\n",
       "  'residue': 'D'},\n",
       " 144: {'start': 584,\n",
       "  'end': 587,\n",
       "  'oligo': 2,\n",
       "  'in_oligo_start': 184,\n",
       "  'in_oligo_end': 187,\n",
       "  'codon': 'GCG',\n",
       "  'residue': 'A'},\n",
       " 145: {'start': 587,\n",
       "  'end': 590,\n",
       "  'oligo': 2,\n",
       "  'in_oligo_start': 187,\n",
       "  'in_oligo_end': 190,\n",
       "  'codon': 'CAG',\n",
       "  'residue': 'Q'},\n",
       " 146: {'start': 590,\n",
       "  'end': 593,\n",
       "  'oligo': 2,\n",
       "  'in_oligo_start': 190,\n",
       "  'in_oligo_end': 193,\n",
       "  'codon': 'AAC',\n",
       "  'residue': 'N'},\n",
       " 147: {'start': 593,\n",
       "  'end': 596,\n",
       "  'oligo': 2,\n",
       "  'in_oligo_start': 193,\n",
       "  'in_oligo_end': 196,\n",
       "  'codon': 'TCT',\n",
       "  'residue': 'S'},\n",
       " 148: {'start': 596,\n",
       "  'end': 599,\n",
       "  'oligo': 2,\n",
       "  'in_oligo_start': 196,\n",
       "  'in_oligo_end': 199,\n",
       "  'codon': 'CAC',\n",
       "  'residue': 'H'},\n",
       " 149: {'start': 599,\n",
       "  'end': 602,\n",
       "  'oligo': 2,\n",
       "  'in_oligo_start': 199,\n",
       "  'in_oligo_end': 202,\n",
       "  'codon': 'AGC',\n",
       "  'residue': 'S'},\n",
       " 150: {'start': 602,\n",
       "  'end': 605,\n",
       "  'oligo': 2,\n",
       "  'in_oligo_start': 202,\n",
       "  'in_oligo_end': 205,\n",
       "  'codon': 'TAT',\n",
       "  'residue': 'Y'},\n",
       " 151: {'start': 605,\n",
       "  'end': 608,\n",
       "  'oligo': 2,\n",
       "  'in_oligo_start': 205,\n",
       "  'in_oligo_end': 208,\n",
       "  'codon': 'TGC',\n",
       "  'residue': 'C'},\n",
       " 152: {'start': 608,\n",
       "  'end': 611,\n",
       "  'oligo': 2,\n",
       "  'in_oligo_start': 208,\n",
       "  'in_oligo_end': 211,\n",
       "  'codon': 'TTT',\n",
       "  'residue': 'F'},\n",
       " 153: {'start': 611,\n",
       "  'end': 614,\n",
       "  'oligo': 3,\n",
       "  'in_oligo_start': 26,\n",
       "  'in_oligo_end': 29,\n",
       "  'codon': 'GAG',\n",
       "  'residue': 'E'},\n",
       " 154: {'start': 614,\n",
       "  'end': 617,\n",
       "  'oligo': 3,\n",
       "  'in_oligo_start': 29,\n",
       "  'in_oligo_end': 32,\n",
       "  'codon': 'ATT',\n",
       "  'residue': 'I'},\n",
       " 155: {'start': 617,\n",
       "  'end': 620,\n",
       "  'oligo': 3,\n",
       "  'in_oligo_start': 32,\n",
       "  'in_oligo_end': 35,\n",
       "  'codon': 'CTG',\n",
       "  'residue': 'L'},\n",
       " 156: {'start': 620,\n",
       "  'end': 623,\n",
       "  'oligo': 3,\n",
       "  'in_oligo_start': 35,\n",
       "  'in_oligo_end': 38,\n",
       "  'codon': 'GAG',\n",
       "  'residue': 'E'},\n",
       " 157: {'start': 623,\n",
       "  'end': 626,\n",
       "  'oligo': 3,\n",
       "  'in_oligo_start': 38,\n",
       "  'in_oligo_end': 41,\n",
       "  'codon': 'CGG',\n",
       "  'residue': 'R'}}"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mlib = MutantLib(lib, \"test\")\n",
    "mlib.annotate_codons()\n",
    "mlib.codons"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
